logserver/src/main/java/alluxio/logserver/AlluxioLogServerProcess.java

```diff
@@ -0,0 +1,240 @@
+ * {@link AlluxioLogServerProcess} allows masters and workers to "push" their logs to a
+ * centralized log server where another copy of the logs will be stored.
+ */
+public class AlluxioLogServerProcess implements Process {
+  private static final String LOGSERVER_CLIENT_LOGGER_APPENDER_NAME = "LOGSERVER_CLIENT_LOGGER";
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioLogServer.class);
+  private static final long STOP_TIMEOUT_MS = 60000;
+  private static final int BASE_SLEEP_TIME_MS = 50;
+  private static final int MAX_SLEEP_TIME_MS = 30000;
+  private static final int MAX_NUM_RETRY = 20;
+
+  private final String mBaseLogsDir;
+  private int mPort;
+  private ServerSocket mServerSocket;
+  private final int mMinNumberOfThreads;
+  private final int mMaxNumberOfThreads;
+  private ExecutorService mThreadPool;
+  private volatile boolean mStopped;
+
+  /**
+   * Construct an {@link AlluxioLogServerProcess} instance.
+   *
+   * @param baseLogsDir base directory to store the logs pushed from remote Alluxio servers
+   */
+  public AlluxioLogServerProcess(String baseLogsDir) {
+    mPort = Configuration.getInt(PropertyKey.LOG_SERVER_PORT);
+    // The log server serves the logging requests from Alluxio workers, Alluxio master, Alluxio
+    // secondary master, and Alluxio proxy. Therefore the number of threads required by
+    // log server is #workers + 1 (master) + 1 (secondary master) + 1 (proxy).
+    mMinNumberOfThreads = Configuration.getInt(PropertyKey.MASTER_WORKER_THREADS_MIN) + 3;
+    mMaxNumberOfThreads = Configuration.getInt(PropertyKey.MASTER_WORKER_THREADS_MAX) + 3;
+    mBaseLogsDir = baseLogsDir;
+    mStopped = true;
+  }
+
+  @Override
+  public void start() throws Exception {
+    startServing();
+  }
+
+  @Override
+  public void stop() throws Exception {
+    stopServing();
+  }
+
+  @Override
+  public void waitForReady() {
+    CommonUtils.waitFor(this + " to start", new Function<Void, Boolean>() {
+      @Override
+      public Boolean apply(Void input) {
+        return mStopped == false;
+      }
+    }, WaitForOptions.defaults().setTimeoutMs(10000));
+  }
+
+  /**
+   * Create and start logging server and client thread pool.
+   */
+  private void startServing() {
+    SynchronousQueue<Runnable> synchronousQueue =
+        new SynchronousQueue<>();
+    mThreadPool =
+        new ThreadPoolExecutor(mMinNumberOfThreads, mMaxNumberOfThreads,
+            STOP_TIMEOUT_MS, TimeUnit.MILLISECONDS, synchronousQueue);
+    try {
+      mServerSocket = new ServerSocket(mPort);
+    } catch (IOException e) {
+      LOG.error("Failed to bind to port {}.", mPort);
+      throw new RuntimeException(e);
+    }
+    mStopped = false;
+    while (!mStopped) {
+      Socket client;
+      try {
+        client = mServerSocket.accept();
+      } catch (IOException e) {
+        LOG.warn("I/O error occured while waiting for connection.");
+        throw new RuntimeException(e);
```

Comment:
is this fatal, or should we `continue`?

Reply:
According to javadoc of `accept`, an `IOException` will be thrown if there is an I/O error while waiting for connection. If there is an I/O error, it may be hard to recover?

Reply:
An I/O error could happen due to a misbehaving client, so it should not kill the server. The javadoc for `ServerSocket#accept` doesn't say that `IOException` will leave the `ServerSocket` in a bad state, so it should be fine to continue listening for connections

Reply:
Done.

Reply:
It is recommended to retry binding the server socket to the same port and handle exception. See the new code.